client/mlConnectionConfigurator.ts

```diff
@@ -0,0 +1,207 @@
+import * as ml from 'marklogic';
+
+import { ClientContext, MlClientParameters } from './marklogicClient';
+
+// This class should NOT use the WorkspaceConfiguration (WC) class imported below.
+// The debuggers don't have access to the vscode namespace, including WC.
+// So if WC is used in the class, the debuggers won't be able to use this class.
+// Note that this also precludes the use of ConfigurationManager in the class.
+export class MlConnectionConfigurator implements ml.ConnectionParams {
+    host: string;
+    port: number;
+    basePath: string;
+    user: string;
+    password: string;
+    database: string;
+    authType: string;
+    ssl: boolean;
+    ca: string;
+    rejectUnauthorized: boolean;
+    // Additional fields not included in ml.ConnectionParams
+    pathToCa: string;
+    modulesDb: string;
+    managePort: number;
+    adminPort: number;
+    testPort: number;
+    manageBasePath: string;
+    adminBasePath: string;
+    restBasePath: string;
+    testBasePath: string;
+
+    constructor(
+        rawParams: Record<string, any>
+    ) {
+        this.host = rawParams.host;
+        this.port = Number(rawParams.port);
+        this.basePath = rawParams.basePath || '';
+        this.managePort = Number(rawParams.managePort);
+        this.manageBasePath = rawParams.manageBasePath || '';
+        this.testPort = Number(rawParams.testPort);
+        this.testBasePath = rawParams.testBasePath || '';
+        this.adminPort = Number(rawParams.adminPort);
+        this.adminBasePath = rawParams.adminBasePath || '';
+        this.user = rawParams.user;
+        this.password = rawParams.password;
+        // contentDb is required for backward compatability
+        this.database = rawParams.contentDb || rawParams.documentsDb || '';
+        this.modulesDb = rawParams.modulesDb || '';
+        this.authType = rawParams.authType;
+        this.ssl = Boolean(rawParams.ssl);
+        this.pathToCa = rawParams.pathToCa || '';
+        this.rejectUnauthorized = Boolean(rawParams.rejectUnauthorized);
+
+        // This check was previously done in the MarklogicClient constructor, but doing
+        // so causes the sameAs function in this class to not behave properly
+        if (this.authType !== 'DIGEST' && this.authType !== 'BASIC') {
+            this.authType = 'DIGEST';
+        }
+    }
+
+    toMlClientParameters(): MlClientParameters {
+        return new MlClientParameters({
+            host: this.host,
+            port: this.port,
+            basePath: this.basePath,
+            user: this.user,
+            pwd: this.password,
+            contentDb: this.database,
+            modulesDb: this.modulesDb,
+            authType: this.authType,
+            ssl: this.ssl,
+            pathToCa: this.pathToCa,
+            rejectUnauthorized: this.rejectUnauthorized,
+        });
+    }
+
+    newMarklogicRestClient(
```

Comment:
This is good and will help readability a lot. I think this suggests the concepts are:

1. There's a "Configuration" of all the settings that a user can provide.
2. There's a "ClientFactory" that you instantiate, giving it a "Configuration".
3. You then call the appropriate "Make new a Client" Method on the "ClientFactory".

Reply:
Exactly what I'm going for.